### ISSUE TITLE
Potential memory leak when using the new ServerStyleSheet

### ISSUE THREAD

**Issue by cdimitroulas** [styled-components/styled-components#866]
## Version
2.0.0

Today I picked up that memory usage was increasing continuously on one of our servers and I tracked it down to a commit which related to upgrading to the latest version of styled-components and editing some server-side code.

It seems that there may be a problem with the `ServerStyleSheet` as we were using `getCSS` before and this problem was not present.

Any help would be much appreciated!

The server-side rendering function which is called looks like this:
```js
function serverSideRender(req, res, store) {
  const preloadedState = store.getState()
  const sheet = new ServerStyleSheet()
  const context = {}

  const markup = renderToString(
    <Provider store={store}>
      <StaticRouter location={req.url} context={context}>
        <StyleSheetManager sheet={sheet.instance}>
          <App />
        </StyleSheetManager>
      </StaticRouter>
    </Provider>,
  )

  const helmet = Helmet.renderStatic()

  res.status(200).send(`
    <!DOCTYPE html>
    <html lang="en">
      <head>
        <meta name="viewport" content="width=device-width, initial-scale=1">
        ${helmet.title.toString()}
        ${helmet.meta.toString()}
        ${helmet.link.toString()}
        <link href="https://fonts.googleapis.com/css?family=Open+Sans:300,400,700" rel="stylesheet">
        <link rel="stylesheet" href="/client.css" />
      </head>
      <body>
        <div id="root">${markup}</div>
        <script>window.__PRELOADED_STATE__ = ${JSON.stringify(preloadedState).replace(/</g, '\\u003c')}</script>
        <script src='/client.js'></script>
      </body>
    </html>
  `)
}
```

**Comment by philpl** [comment]
The code snippet is unfortunately not that helpful :/ It'd be good to know how many pages you have, and whether memory usage is increasing over time, when only the same pages are being hit, and additionally we'd need to establish that this is caused by the root StyleSheet, since that's the one that collects global styles, which is thus the only instance that can hold residual rules after the ServerStyleSheet that was created is GCd.

**Comment by cdimitroulas** [comment]
We will do some tests next week to see whether the StylesheetProvider or the ServerStyleSheet is causing the issue.
Our web app is receiving very minimal requests right now as we are still developing it - the total number of pages is around 10 at the moment.
The memory usage is increasing at a constant rate regardless of the number of requests as far as I can tell.

I will feedback more information as we investigate further on Monday.

**Comment by philpl** [comment]
@cdimitroulas if the memory usage is increasing at a constant rate, and is not correlated to the number of requests e.g. rendering the app on the server, that's a good sign, that it's actually not responsible :)

**Comment by alerosa** [comment]
@philpl we did come other tests and the memory usage seems  correlated with the number of requests.
We have the problem if we use the `StyleSheetManager` or the `sheet.collectStyles` method so for now we reverted back using ` sheet.getStyleTags()`  and then manually put the style tags inside the template string.

Here the code snippets that we tested as they may could be useful as reference:

Case 1 - StyleSheetManager (**memory leak**)
```jsx
function serverSideRender(req, res, store) {
  const preloadedState = store.getState()
  const sheet = new ServerStyleSheet()

  const markup = renderToString(
    <Provider store={store}>
      <StaticRouter location={req.url}>
        <StyleSheetManager sheet={sheet.instance}>
          <App />
        </StyleSheetManager>
      </StaticRouter>
    </Provider>,
  )

  res.status(200).send(`
    <!DOCTYPE html>
    <html lang="en">
      <head>
      </head>
      <body>
        <div id="root">${markup}</div>
      </body>
    </html>
 `)
}
```

Case 2 - sheet.collectStyles (**memory leak**)
```jsx
function serverSideRender(req, res, store) {
  const preloadedState = store.getState()
  const sheet = new ServerStyleSheet()

  const AppWrapper = () => (
    <Provider store={store}>
      <StaticRouter location={req.url}>
        <App />
      </StaticRouter>
    </Provider>
  )

  const markup = renderToString(sheet.collectStyles(<AppWrapper />))

  res.status(200).send(`
    <!DOCTYPE html>
    <html lang="en">
      <head>
      </head>
      <body>
        <div id="root">${markup}</div>
      </body>
    </html>
 `)
}
```

Case 3 - sheet.getStyleTags (**works just fine**)
```jsx
function serverSideRender(req, res, store) {
  const preloadedState = store.getState()
  const sheet = new ServerStyleSheet()
  const css = sheet.getStyleTags()

  const markup = renderToString(
    <Provider store={store}>
      <StaticRouter location={req.url}>
        <App />
      </StaticRouter>
    </Provider>,
  )

  res.status(200).send(`
    <!DOCTYPE html>
    <html lang="en">
      <head>
        ${css}
      </head>
      <body>
        <div id="root">${markup}</div>
      </body>
    </html>
 `)
}
```

**Comment by cdimitroulas** [comment]
any update on this?

**Comment by DeividasK** [comment]
@philpl Any update on this? We are experience this as well. The note here about the accumulation of styles is what is affecting us - https://www.styled-components.com/docs/advanced#server-side-rendering. We can't do aggressive caching, because there's a lot of moving parts on our site, so any idea on when this would be fixed? Is anyone working on this or can I take a stab at this myself?

**Comment by philpl** [comment]
@DeividasK If you're affected by the next.js style accumulation, then I'd recommend you to modify its code and vendor it. That's what I've now done on our website: https://github.com/styled-components/styled-components-website/commit/f9a93201c4dbb44545c7c87cc8f551d3c73b7958

**Comment by DeividasK** [comment]
@philpl We are affected by the style accumulation, but we are not using next.js - we are using a modified version of react-boilerplate. Is there a way to flush styles after I request for styles? Or before styles are collect? Is there a reason why this wasn't implemented initially?

**Comment by philpl** [comment]
@DeividasK If the styles accumulate for you this is not the right issue to post in. Please refer to the docs on how to implement SSR correctly. In this case the styles won't accumulate: https://www.styled-components.com/docs/advanced#server-side-rendering

Also use the babel-plugin's SSR option to keep component ids stable: https://www.styled-components.com/docs/advanced#serverside-rendering

**Comment by geelen** [comment]
I've been away on holiday so I'm just now looking back over the issue tracker, so I'm not fully across this but if there is a problem I'm keen to get it cleared up.

For starters, though, the correct usage includes _both_ calling `collectStyles` and `getStyleTags` (see the docs), so I don't understand @alerosa's code snippets. And it's designed in such a way as to not accumulate styles from render to render. There's a small chance that if you're _not_ using the babel transform some memory might accumulate but... not much. And really SSR requires the transform to be 100% deterministic anyway so you should have that running.

Basically, the styles get serialised and detached from the main SC instance when `getStyleTags` is called. Then once your `sheet` falls out of scope the GC should clean up it and the stylesheet it still has reference to. So I think the best way to solve this is to post your code at the point that it gets rendered and we can figure out if something else is hanging on to a reference it shouldn't be.

**Comment by DeividasK** [comment]
@philpl @geelen you are correct. This particular issue happened to me, because I was using babel-plugin-styled-components only for the main repo, but not for the package that is shared between different repos. Adding the plugin to the shared package fixed my issue. Thanks for the help!

**Comment by geelen** [comment]
Glad to hear it!

**Comment by cdimitroulas** [comment]
@geelen thanks for your comment - it seems we misunderstood the correct use of the StyleSheetManager and that was what caused the memory leak.
That has been corrected now and everything seems to be working properly :+1:

**Comment by alerosa** [comment]
@geelen yeah, thanks a lot for your support!

**Comment by rovansteen** [comment]
@DeividasK running in to a similar issue, but I am a bit confused why that would be an issue. You don't always have control over what you import so if you would use a third party styled components library that does not have the SSR babel plugin you would have a memory leak if you try to render it server side? @geelen @philpl is that correct?
